offchain-modules/lib/src/relay/eth_relay.rs

```diff
@@ -177,51 +181,82 @@ impl ETHRelayer {
             info!("waiting for new eth header. tip_header_number: {}, latest_submit_header_number: {}", tip_header_number, latest_submit_header_number);
             return Ok(latest_submit_header_number);
         }
-        // sync cached blocks
-        let start = if tip_header_number > MAIN_HEADER_CACHE_LIMIT as u64 {
-            tip_header_number - MAIN_HEADER_CACHE_LIMIT as u64
-        } else {
-            1
+
+        let force_config = ForceConfig::new(self.config_path.as_str())?;
+        let db_path = force_config.rocksdb_path;
+        // make tx
+        let cell =
+            get_live_cell_by_typescript(&mut self.generator.indexer_client, cell_script.clone())
+                .map_err(|err| anyhow::anyhow!(err))?
+                .ok_or_else(|| anyhow::anyhow!("no cell found"))?;
+
+        let last_cell_output_data = cell.output_data.as_bytes();
+
+        let mut last_cell_latest_height = 0u64;
+
+        let (start_height, mut smt_tree) = match last_cell_output_data.len() {
+            0 => {
+                let rocksdb_store = rocksdb::RocksDBStore::new(db_path.clone());
+                (
+                    tip_header_number,
+                    rocksdb::SMT::new(sparse_merkle_tree::H256::zero(), rocksdb_store),
+                )
+            }
+            _ => {
+                let (start_height, latest_height, merkle_root) =
+                    parse_merkle_cell_data(last_cell_output_data.to_vec())?;
+                last_cell_latest_height = latest_height;
+                let rocksdb_store = rocksdb::RocksDBStore::open(db_path.clone());
+                (
+                    start_height,
+                    rocksdb::SMT::new(merkle_root.into(), rocksdb_store),
+                )
+            }
         };
+
+        // 1. if first time update eth light client, only update latest header
+        // 2. if last cell latest height not exceed CONFIRM, update block from 1 to latest
+        // 3. if last cell latest height over CONFIRM, update block from last_cell_latest_height - CONFIRM to latest
```

Comment:
The logic here is wrong. We should start from latest chain block number, backtrace the db to latest matched block hash, start from there.

The meaning of merkle root in light client cell data is always the root of all k-v from start_height to latest_height in data. The SMT now have holes(because we skip some keys), other offchain process can not generate the SMT.